### PR TITLE
Add linalg.lu

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73878
* #73877
* #74046
* #73806
* #73804
* #73803
* #72935
* __->__ #67833

This PR modifies `lu_unpack` by:
- Using less memory when unpacking `L` and `U`
- Fuse the subtraction by `-1` with `unpack_pivots_stub`
- Define tensors of the correct types to avoid copies
- Port `lu_unpack` to be a strucutred kernel so that its `_out` version does not incur on extra copies

Then we implement `linalg.lu` as a structured kernel, as we want to
compute its derivative manually. We do so because composing the derivatives
of `torch.lu_factor` and `torch.lu_unpack` would be less efficient.

This new function and `lu_unpack` comes with all the things it can come:
forward and backward ad, decent docs, correctness tests, OpInfo, complex support,
support for metatensors and support for vmap and vmap over the gradients.

I really hope we don't continue adding more features.

This PR also avoids saving some of the tensors that were previously
saved unnecessarily for the backward in `lu_factor_ex_backward` and
`lu_backward` and does some other general improvements here and there
to the forward and backward AD formulae of other related functions.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

**Comment by Lezcano** [comment]
@IvanYashchuk any idea why ROCM is failing here:
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.3.1-py3.6-trigger/14118/
?

**Comment by IvanYashchuk** [comment]
Because it's not allowed to use __restrict__ with auto.
> /var/lib/jenkins/workspace/aten/src/ATen/native/hip/LinearAlgebra.hip:125:16: error: restrict requires a pointer or reference ('auto' is invalid)
    const auto __restrict__ unpacked_pivots_data = reinterpret_cast<int64_t*>(unpacked_pivots_ptr + offsets[0]);
               ^
/var/lib/jenkins/workspace/aten/src/ATen/native/hip/LinearAlgebra.hip:126:16: error: restrict requires a pointer or reference ('auto' is invalid)
    const auto __restrict__ pivots_data = reinterpret_cast<const int32_t*>(pivots_ptr + offsets[1]);
               ^
2 warnings and 2 errors generated when compiling for gfx900.

**Comment by Lezcano** [comment]
Thank you, completely missed that one!

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1682,6 +1744,103 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info(const Tensor& self, bool comput
...
+  const auto m = A.sizes().end()[-2];
+  const auto n = A.sizes().end()[-1];
+
+  // A.shape[-2:] == (m, n)
+  // P.shape[-2:] == (m, m)
+  // L.shape[-2:] == (m, k)
+  // U.shape[-2:] == (k, n)
+  // with k = min(m, n)
+
+  // Use L as it has the correct size
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1765]: L is used for storing the result of lu_factor because it needs an m x n matrix to save the result, right?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1682,6 +1744,103 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info(const Tensor& self, bool comput
...
+      L.diagonal(0, -2, -1).fill_(1.);
+    } else {
+      // The order of triu and tril is important as we may have LU.is_same(U)
+      at::tril_out(const_cast<Tensor&>(L), m == n ? LU : LU.narrow(-1, 0, m), -1);
+      L.diagonal(0, -2, -1).fill_(1.);
+      at::triu_out(const_cast<Tensor&>(U), LU, 0);
+    }
+  }
+  if (unpack_pivots) {
+    // lu_factor_ex returns an int32 1-based indexing, which is what we have in `info`
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1816]:
```suggestion
    // lu_factor_ex returns an int32 1-based indexing, which is what we have in `pivots`
```

File: aten/src/ATen/native/cuda/LinearAlgebra.cu [hunk 3]
```diff
@@ -1,7 +1,7 @@
-#define TORCH_ASSERT_NO_OPERATORS
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/LinearAlgebra.cu:1]: Why is this removed?

File: aten/src/ATen/native/LinearAlgebra.h [hunk 4]
```diff
@@ -18,12 +18,4 @@ DECLARE_DISPATCH(addr_fn, addr_stub);
 
 using linalg_vector_norm_fn = void(*)(TensorIterator &, Scalar);
 DECLARE_DISPATCH(linalg_vector_norm_fn, linalg_vector_norm_stub);
-
-using unpack_pivots_fn = void(*)(
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.h:22]: The name of the file is unfortunate but BatchLinearAlgebra files, where this function declaration was moved, are mostly for LAPACK&BLAS wrappers. I'd keep `unpack_pivots_stub` in this file and have the git diff shorter.

I think it's also fine to leave it as it is and move the declarations if that's preferred.

**Review by IvanYashchuk — APPROVED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1765]: Yup!

File: aten/src/ATen/native/cuda/LinearAlgebra.cu (hunk 3 above)
Comment (Lezcano) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:1]: Because we added the dependency with `BatchLinearAlgebra.h` down there.

This one was tricky, as we don't have a file to have our custom kernels for linalg. This is, by the looks of it, the only linalg function where we use a TensorIterator. It's no big deal, as this file is super short, so removing that macro from here is fine really.

File: aten/src/ATen/native/LinearAlgebra.h (hunk 4 above)
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebra.h:22]: I might be wrong here, but imo these two files are just a legacy split which should not exist anymore. Both of them have LAPACK/BLAS wrappers. I just moved this function to be close to the other LU functions within `BatchedLinearAlgebra` as they are logically related.

**Comment by nikitaved** [comment]
Hey, @Lezcano , I will also have a pass today/tomorrow.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -1660,6 +1722,103 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info(const Tensor& self, bool comput
...
+                        .expand(perm_sizes)
+                        .contiguous();
+
+    auto iter = TensorIteratorConfig()
+      .set_check_mem_overlap(false)
+      .check_all_same_dtype(false)
+      .resize_outputs(false)
+      .declare_static_shape(pivots.sizes(), /*squash_dim=*/pivots.dim() - 1)
+      .add_output(perm)
+      .add_owned_input(pivots.contiguous())
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1809]: Did not see that before. What does it do?

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1809]: I'm passing an r-value there, so I'm asking TensorIterator to keep this `Tensor` for the duration of TI's life.

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1809]: From a comment in TI's header file:
>  // Stores input/output Tensors while incrementing the reference count.
  // Note that add_{in,out}put are nearly always what you
  // want, and the exception (adding an unnamed temporary) won't
  // compile.

File: aten/src/ATen/native/cuda/LinearAlgebra.cu [hunk 6]
```diff
@@ -100,56 +100,38 @@ static void _launch_kernel(int total_n_elems, func_t f) {
...
-  auto loop = [=]C10_DEVICE(int i) {
-    auto offsets = offset_calculator.get(i);
+  auto loop = [=]C10_DEVICE(const int idx) {
+    const auto offsets = offset_calculator.get(idx);
 
-    auto* unpacked_pivots_data = reinterpret_cast<int32_t*>(
-      unpacked_pivots_ptr + offsets[0]);
-    const auto* const __restrict__ pivots_data = reinterpret_cast<const int32_t*>(
-      pivots_ptr + offsets[1]);
+    int64_t* const __restrict__ perm_data = reinterpret_cast<int64_t*>(perm_ptr + offsets[0]);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/LinearAlgebra.cu:119]: Not sure that restrict works here. `perm_data` is the pointer that you write into, and restrict is to be used with pointers that just refer to data.

File: aten/src/ATen/native/cuda/LinearAlgebra.cu [hunk 7]
```diff
@@ -100,56 +100,38 @@ static void _launch_kernel(int total_n_elems, func_t f) {
...
 
-  char* unpacked_pivots_ptr = reinterpret_cast<char*>(iter.data_ptr(0));
-  const char* const __restrict__ pivots_ptr = reinterpret_cast<const char*>(iter.data_ptr(1));
+  const auto perm_ptr = reinterpret_cast<char*>(iter.data_ptr(0));
+  const auto pivots_ptr = reinterpret_cast<const char*>(iter.data_ptr(1));
 
-  auto loop = [=]C10_DEVICE(int i) {
-    auto offsets = offset_calculator.get(i);
+  auto loop = [=]C10_DEVICE(const int idx) {
+    const auto offsets = offset_calculator.get(idx);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/LinearAlgebra.cu:117]: You are obsessed with `const` :)

File: aten/src/ATen/native/cuda/LinearAlgebra.cu [hunk 8]
```diff
@@ -100,56 +100,38 @@ static void _launch_kernel(int total_n_elems, func_t f) {
...
-      pivots_ptr + offsets[1]);
+    int64_t* const __restrict__ perm_data = reinterpret_cast<int64_t*>(perm_ptr + offsets[0]);
+    const int32_t* const __restrict__ pivots_data = reinterpret_cast<const int32_t*>(pivots_ptr + offsets[1]);
 
     // QUESTION: can we mix 64bit offsets with 32bit Iterator indexing?
     for (int64_t i = 0; i < dim_size; ++i) {
       thrust::swap(
-        unpacked_pivots_data[i],
-        unpacked_pivots_data[pivots_data[i]]
+        perm_data[i],
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/LinearAlgebra.cu:125]: Maybe we should add a comment that it is expected that `perm_data` is `kLong`, while `pivots_data` is 1-based `kInt` indices...

File: aten/src/ATen/native/cuda/LinearAlgebra.cu (hunk 7 above)
Comment (Lezcano) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:117]: That is not false :)
I mean, you may get free optimisations / warnings from the compiiler so why not?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 9]
```diff
@@ -4413,35 +4413,50 @@ Tensor lu_solve_jvp(
...
 ) {
-  auto L_grad = grads[1];
-  auto U_grad = grads[2];
-
-  auto m = LU_data.size(-2);
-  auto n = LU_data.size(-1);
-  auto k = std::min(m, n);
-
-  TORCH_CHECK(unpack_data, "lu_unpack_backward: cannot compute gradients unless unpack_data=True");
+  if (!L_grad.defined() && !U_grad.defined()) {
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:4421]: logic and is necessary here and cannot be replaced with or?

File: aten/src/ATen/native/cuda/LinearAlgebra.cu (hunk 6 above)
Comment (Lezcano) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:119]: `__restrict__` does assure the compiler that this pointer does not alias with any other pointer, which is true. Now, it is true that it does not qualify for the const-optimisation that CUDA sometimes performs because the data is itself not const.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 9 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:4421]: What do you mean here? If both gradients are zero, then the input gradient is zero, but we cannot say much more otherwise.

**Review by nikitaved — APPROVED** [review]
LGTM!

File: tools/autograd/derivatives.yaml [hunk 10]
```diff
@@ -936,17 +936,28 @@
...
 - name: linalg_lu_factor_ex(Tensor A, *, bool pivot=True, bool check_errors=False) -> (Tensor LU, Tensor pivots, Tensor info)
-  A: lu_factor_ex_backward(grad, A, LU, pivots)
-  LU: lu_factor_ex_jvp(A_t, LU, pivots)
+  A: lu_factor_ex_backward(grad, LU, pivots, pivot)
+  LU: lu_factor_ex_jvp(A_t, LU, pivots, pivot)
   output_differentiability: [True, False, False]
 
+- name: linalg_lu(Tensor A, *, bool pivot=True) -> (Tensor P, Tensor L, Tensor U)
+  A: linalg_lu_backward(grad_L, grad_U, P, L, U, pivot)
+  P: std::get<1>(std::make_tuple(A_t, zeros_like(P))) # WRONG, this should be non_differentiable
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply tools/autograd/derivatives.yaml:945]: Actually, I believe that this is not necessary if I simply filter out P in the OpInfo. I forgot to do it, thanks for pointing this out.

**Comment by Lezcano** [comment]
The reds in the CI are unrelated.

**Comment by Lezcano** [comment]
@mruberry could I have this stamped please?

**Review by mruberry — APPROVED** [review]
Cool! Stamped!

**Comment by malfet** [comment]
This PR broke internal builds, because BatchLinearAlgebraKernels.cpp references `for_each` method from TensorIterator.cpp, but for some reason they are part of different filelists, see :
https://github.com/pytorch/pytorch/blob/6c615a21a0d6a01f0b440b483bdd26ce5ad204ea/tools/build_variables.bzl#L1377
but
https://github.com/pytorch/pytorch/blob/6c615a21a0d6a01f0b440b483bdd26ce5ad204ea/tools/build_variables.bzl#L1065

@cccclai  you've introduced the split, do you know why some native files are in one filelist, but others in another?
